internal/jimm/cloud.go

```diff
@@ -20,35 +20,173 @@ import (
 	ofganames "github.com/canonical/jimm/v3/internal/openfga/names"
 )
 
+// GetUserCloudAccess returns users access level for the specified cloud.
+func (j *JIMM) GetUserCloudAccess(ctx context.Context, user *openfga.User, cloud names.CloudTag) (string, error) {
+	accessLevel := user.GetCloudAccess(ctx, cloud)
+	return permissions.ToCloudAccessString(accessLevel), nil
+}
+
+// GetCloud retrieves the cloud for the given cloud tag. If the cloud
+// cannot be found then an error with the code CodeNotFound is
+// returned. If the user does not have permission to view the cloud then an
+// error with a code of CodeUnauthorized is returned. If the user only has
+// add-model access to the cloud then the returned Users field will only
+// contain the authentcated user.
+func (j *JIMM) GetCloud(ctx context.Context, user *openfga.User, tag names.CloudTag) (jujuparams.Cloud, error) {
+	const op = errors.Op("jimm.CloudInfo")
+	zapctx.Info(ctx, string(op))
+	var cl dbmodel.Cloud
+	cl.SetTag(tag)
+
+	if err := j.Database.GetCloud(ctx, &cl); err != nil {
+		return jujuparams.Cloud{}, errors.E(op, err)
+	}
+	// TODO (SimoneDutto): refactor this to use `user.IsCloudAdmin()`
+	accessLevel, err := j.GetUserCloudAccess(ctx, user, tag)
+	if err != nil {
+		return jujuparams.Cloud{}, errors.E(op, err)
+	}
+	if accessLevel == "" {
+		return jujuparams.Cloud{}, errors.E(op, errors.CodeUnauthorized, "unauthorized")
+	}
+
+	return j.getCloudFromController(ctx, cl)
+}
+
+func (j *JIMM) getCloudFromController(ctx context.Context, cl dbmodel.Cloud) (jujuparams.Cloud, error) {
+	const op = errors.Op("jimm.getCloudFromController")
+
+	var jCl jujuparams.Cloud
+	controllers := make([]dbmodel.Controller, 0)
+	for _, cr := range cl.Regions {
+		for _, c := range cr.Controllers {
+			controllers = append(controllers, c.Controller)
+		}
+	}
+	err := j.firstSuccessfulController(ctx, controllers, func(api API) error {
+		return api.Cloud(ctx, cl.ResourceTag(), &jCl)
+	})
+	if err != nil {
+		return jCl, errors.E(op, err)
+	}
+	jCl.IsControllerCloud = false // jimm doesn't know where is deployed, so none of the clouds should have this field set.
+	return jCl, nil
+}
+
+func (j *JIMM) GetClouds(ctx context.Context, user *openfga.User) (map[string]jujuparams.Cloud, error) {
+	const op = errors.Op("jimm.GetClouds")
+	zapctx.Info(ctx, string(op))
+
+	var clouds []dbmodel.Cloud
+
+	err := j.ForEachUserCloud(ctx, user, func(c *dbmodel.Cloud) error {
+		clouds = append(clouds, *c)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	results := make(map[string]jujuparams.Cloud, 0)
+	for _, cl := range clouds {
+		cloud, err := j.getCloudFromController(ctx, cl)
+		if err != nil {
+			return nil, errors.E(op, err)
+		}
+		results[cl.ResourceTag().String()] = cloud
+	}
```

Comment:
I have a concern that this will be annoyingly slow. Not so slow that it's bad but since getting a cloud requires a sequential call to a controller... I've experienced running `juju <command>` to take several seconds sometimes so I hope we don't have that issue here.
Some caching in the future will of course help and then be similar in a way to what the database was doing.